components/odh-notebook-controller/controllers/notebook_rbac.go

```diff
@@ -0,0 +1,166 @@
+		// Check ClusterRole if roleRefKind is ClusterRole
+		clusterRole := &rbacv1.ClusterRole{}
+		err := r.Get(ctx, types.NamespacedName{Name: roleRefName}, clusterRole)
+		if err != nil {
+			if apierrs.IsNotFound(err) {
+				// ClusterRole not found
+				return false, nil
+			}
+			return false, err // Some other error occurred
+		}
+	} else {
+		// Check Role if roleRefKind is Role
+		role := &rbacv1.Role{}
+		err := r.Get(ctx, types.NamespacedName{Name: roleRefName, Namespace: namespace}, role)
+		if err != nil {
+			if apierrs.IsNotFound(err) {
+				// Role not found
+				return false, nil
+			}
+			return false, err // Some other error occurred
+		}
+	}
+	return true, nil // Role or ClusterRole exists
+}
+
+// Helper function to delete the RoleBinding if the notebooks deleted
+func (r *OpenshiftNotebookReconciler) deleteRoleBinding(ctx context.Context, rolebindingName, namespace string) error {
+	roleBinding := &rbacv1.RoleBinding{}
+	err := r.Get(ctx, types.NamespacedName{Name: rolebindingName, Namespace: namespace}, roleBinding)
+	if err != nil {
+		if apierrs.IsNotFound(err) {
+			return nil // RoleBinding not found, nothing to delete
+		}
+		return err // Some other error occurred
+	}
+
+	// Delete the RoleBinding
+	return r.Delete(ctx, roleBinding)
+}
+
+// reconcileRoleBinding manages creation, update, and deletion of RoleBindings and ClusterRoleBindings
+func (r *OpenshiftNotebookReconciler) reconcileRoleBinding(
+	notebook *nbv1.Notebook, ctx context.Context, rolebindingName, roleRefKind, roleRefName string) error {
+
+	log := r.Log.WithValues("notebook", types.NamespacedName{Name: notebook.Name, Namespace: notebook.Namespace})
+
+	// Check if the Role or ClusterRole exists before proceeding
+	roleExists, err := r.checkRoleExists(ctx, roleRefKind, roleRefName, notebook.Namespace)
+	if err != nil {
+		log.Error(err, "Error checking if Role exists", "Role.Kind", roleRefKind, "Role.Name", roleRefName)
+		return err
+	}
+	if !roleExists {
+		return nil // Skip if dspa Role is not found on the namespace
+	}
+
+	// Create a new RoleBinding based on provided parameters
+	roleBinding := NewRoleBinding(notebook, rolebindingName, roleRefKind, roleRefName)
+
+	// Check if the RoleBinding already exists
+	found := &rbacv1.RoleBinding{}
+	err = r.Get(ctx, types.NamespacedName{Name: rolebindingName, Namespace: notebook.Namespace}, found)
+	if err != nil && apierrs.IsNotFound(err) {
+		log.Info("Creating RoleBinding", "RoleBinding.Namespace", roleBinding.Namespace, "RoleBinding.Name", roleBinding.Name)
+		err = r.Create(ctx, roleBinding)
+		if err != nil {
+			log.Error(err, "Failed to create RoleBinding", "RoleBinding.Namespace", roleBinding.Namespace, "RoleBinding.Name", roleBinding.Name)
+			return err
+		}
+		return nil
+	} else if err != nil {
+		log.Error(err, "Failed to get RoleBinding")
+		return err
+	}
+
+	// Update RoleBinding if the subjects differ
+	if !reflect.DeepEqual(roleBinding.Subjects, found.Subjects) {
```

Comment:
`DeepEqual` is not the best function to use here, but I don't mind since we do use it everywhere already.

* https://github.com/opendatahub-io/kubeflow/issues/427

and in kubeflow

* https://github.com/kubeflow/notebooks/issues/69

Reply:
Hmm probably we can refactor the code base to use `equality.Semantic.DeepEqual`, but in the whole project. 
But we can kick off a separated task for this to refine it and prioritize it.

Reply:
But, should you keep using it now that we know it's wrong? Maybe in this PR yes, but what about the next PR?

We should certainly _not_ refactor this in our fork and leaving upstream kubeflow unchanged. Kubeflow do, we do.